plugins/flatpak/gs-flatpak.c

```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
 	return TRUE;
 }
 
+/*
+ * gs_flatpak_get_services_app_for_runtime:
+ * Specific to Endless runtimes, gets a GsApp reference for the related
+ * "EknServices" app that provides system services for apps that use the
+ * given runtime.
+ *
+ * Returns NULL if no services app was required or there was an error.
+ * Does not take a GError since any error should be recoverable. However, in
+ * order to make sure the operation was not cancelled, you should check the
+ * status of the GCancellable after calling this.
+ */
+static GsApp *
+gs_flatpak_get_services_app_for_runtime (GsFlatpak *self, GsApp *runtime,
+                                         GCancellable *cancellable)
+{
+	const gchar *runtime_id, *runtime_branch, *services_branch, *arch;
+	g_autofree gchar *services_id = NULL;
+	g_autofree gchar *description = NULL;
+	g_autoptr(FlatpakRef) services_ref = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GsApp) retval = NULL;
+
+	runtime_id = gs_app_get_id (runtime);
+	runtime_branch = gs_app_get_branch (runtime);
+
+	if ((g_strcmp0 (runtime_id, "com.endlessm.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "eos3.1") == 0) ||
+	    (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "1") == 0)) {
+		services_id = g_strdup ("com.endlessm.EknServices");
+		services_branch = "eos3";
+	} else if (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0) {
+		if (g_strcmp0 (runtime_branch, "master") == 0)
+			services_id = g_strdup ("com.endlessm.EknServices2");
+		else
+			services_id = g_strdup_printf ("com.endlessm.EknServices%s",
+			                               runtime_branch);
+		services_branch = "stable";
+	} else {
+		/* Runtime doesn't require an EknServices app */
+		return NULL;
+	}
+
+	/* Construct a GsApp for the EknServices we determined we needed */
+	arch = gs_flatpak_app_get_ref_arch (runtime);
+	description = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	services_ref = flatpak_ref_parse (description, &error);
+	if (!services_ref) {
```

Comment:
Richard doesn't like the `!pointer` notation (prefers `!= NULL`). So even though I prefer to do it as you did, I try to keep the coding style in core modules close to upstream (even if this will never be merged). Small nitpick though.